src/D3MHub.sol

```diff
@@ -221,121 +219,121 @@ contract D3MHub {
+        if (currentAssets > ink) { // If fees were generated
+            uint256 fixInk = _min(
+                currentAssets - ink, // fees generated
+                ink < lineWad // if previously CDP was under debt ceiling
+                    ? lineWad + maxWithdraw - ink // we can just go above debt ceiling + maxWithdraw
+                    : maxWithdraw // we can just go above current state + maxWithdraw
+            );
+            require(fixInk <= MAXINT256, "D3MHub/overflow");
+            vat.slip(ilk, address(_pool), int256(fixInk)); // Generate extra collateral
+            vat.frob(ilk, address(_pool), address(_pool), address(this), int256(fixInk), 0); // Lock it
+            ink += fixInk;
+            emit Fees(ilk, fixInk);
+        }
+        if (art < ink) { // If there was permissionless DAI paid or fees added as collateral
+            address _vow = vow;
+            uint256 fixArt = ink - art; // Amount of fees + permissionless DAI paid we will now transform to debt
+            art += fixArt;
+            require(art <= MAXINT256, "D3MHub/overflow");
+            vat.suck(_vow, _vow, fixArt * RAY); // This needs to be done to make sure we can deduct sin[vow] and vice in the next call
+            vat.grab(ilk, address(_pool), address(_pool), _vow, 0, int256(fixArt)); // Generating the debt
         }
 
-        // To save gas you can bring the fees back with the unwind
-        uint256 total = amount + fees;
-        _pool.withdraw(total);
-        daiJoin.join(address(this), total);
+        // Determine if it needs to unwind or wind
 
-        // normalized debt == erc20 DAI to pool (Vat rate for D3M ilks fixed to 1 RAY)
+        uint256 Line = vat.Line();
+        uint256 debt = vat.debt();
+        uint256 targetAssets;
+        uint256 toUnwind;
 
-        if (mode == Mode.NORMAL) {
-            vat.frob(ilk, address(_pool), address(_pool), address(this), -int256(amount), -int256(amount));
-            vat.slip(ilk, address(_pool), -int256(amount));
-            vat.move(address(this), vow, fees * RAY);
-        } else if (mode == Mode.D3M_CULLED) {
-            vat.slip(ilk, address(_pool), -int256(amount));
-            vat.move(address(this), vow, total * RAY);
+        if (ilks[ilk].tic != 0 || !ilks[ilk].plan.active()) { // If D3M is caged (but not culled) or plan is not active
+            toUnwind = type(uint256).max; // We make sure to enter the unwind path
         } else {
-            // This can be done with the assumption that the price of 1 collateral unit equals 1 DAI.
-            // That way we know that the prev End.skim call kept its gap[ilk] emptied as the CDP was always collateralized.
-            // Otherwise we couldn't just simply take away the collateral from the End module as the next line will be doing.
-            vat.slip(ilk, address(_end), -int256(amount));
-            vat.move(address(this), vow, total * RAY);
+            if (art > lineWad) {
```

Comment:
So about the lack of symmetry, I guess there are 3 things that bothered me, and probably not much we want to change about them:

1. The winding check is more compact - we just do _min(_min(_min.. since we already know all the substractions are safe.
2. There is no check of `if (toWind > 0)` like we have for `toUnWind`, since we go to the wind case even if the amount is 0 (so we can emit an event).
3. `getTargetAssets` is called as part of the unwind checks although it is relevant to both unwind and wind.

(1) and (2) are fine, regarding (3), you might say it's completely redundant but we can do:

```
		if (ilks[ilk].tic != 0 || !ilks[ilk].plan.active()) { // If D3M is caged (but not culled) or plan is not active
			toUnwind = type(uint256).max; // We make sure to enter the unwind path
		} else {
			targetAssets = ilks[ilk].plan.getTargetAssets(currentAssets);
		}

		if (toUnwind == 0) {
			if (art > lineWad) {
....
```

Not insisting on it in any case.

Reply:
Made the change discussed in the call. If feels good this way we can get this comment solved.